src/ynEIGEN/WithdrawalsProcessor.sol

```diff
@@ -0,0 +1,432 @@
+// SPDX-License-Identifier: BSD 3-Clause License
+pragma solidity ^0.8.24;
+
+import {IERC20} from "lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import {IERC4626} from "lib/openzeppelin-contracts/contracts/interfaces/IERC4626.sol";
+import {AccessControlUpgradeable} from
+    "lib/openzeppelin-contracts-upgradeable/contracts/access/AccessControlUpgradeable.sol";
+import {Initializable} from "lib/openzeppelin-contracts-upgradeable/contracts/proxy/utils/Initializable.sol";
+
+import {IDelegationManager} from "lib/eigenlayer-contracts/src/contracts/interfaces/IDelegationManager.sol";
+import {IStrategy} from "lib/eigenlayer-contracts/src/contracts/interfaces/IStrategy.sol";
+
+import {IwstETH} from "../external/lido/IwstETH.sol";
+
+import {IWithdrawalQueueManager} from "../interfaces/IWithdrawalQueueManager.sol";
+import {ITokenStakingNodesManager} from "../interfaces/ITokenStakingNodesManager.sol";
+import {ITokenStakingNode} from "../interfaces/ITokenStakingNode.sol";
+import {IAssetRegistry} from "../interfaces/IAssetRegistry.sol";
+import {IYieldNestStrategyManager} from "../interfaces/IYieldNestStrategyManager.sol";
+import {IynEigen} from "../interfaces/IynEigen.sol";
+import {IRedemptionAssetsVault} from "../interfaces/IRedemptionAssetsVault.sol";
+import {IWrapper} from "../interfaces/IWrapper.sol";
+import {IWithdrawalsProcessor} from "../interfaces/IWithdrawalsProcessor.sol";
+
+contract WithdrawalsProcessor is IWithdrawalsProcessor, Initializable, AccessControlUpgradeable {
+
+    uint256 public totalQueuedWithdrawals;
+
+    uint256 public minPendingWithdrawalRequestAmount;
+
+    IDs private _ids;
+
+    mapping(uint256 id => QueuedWithdrawal) private _queuedWithdrawals;
+    mapping(uint256 fromId => uint256 toId) public batch;
+
+    // yieldnest
+    IWithdrawalQueueManager public immutable withdrawalQueueManager;
+    ITokenStakingNodesManager public immutable tokenStakingNodesManager;
+    IAssetRegistry public immutable assetRegistry;
+    IYieldNestStrategyManager public immutable ynStrategyManager;
+    IynEigen public immutable yneigen;
+    IRedemptionAssetsVault public immutable redemptionAssetsVault;
+    IWrapper public immutable wrapper;
+
+    // eigenlayer
+    IDelegationManager public immutable delegationManager;
+
+    // assets
+    IERC20 private constant STETH = IERC20(0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84);
+    IwstETH private constant WSTETH = IwstETH(0x7f39C581F595B53c5cb19bD0b3f8dA6c935E2Ca0);
+    IERC4626 private constant WOETH = IERC4626(0xDcEe70654261AF21C44c093C300eD3Bb97b78192);
+    IERC20 private constant OETH = IERC20(0x856c4Efb76C1D1AE02e20CEB03A2A6a08b0b8dC3);
+
+    // used to prevent rounding errors
+    uint256 private constant MIN_DELTA = 1000;
+
+    // roles
+    bytes32 public constant KEEPER_ROLE = keccak256("KEEPER_ROLE");
+
+    //
+    // Constructor
+    //
+    constructor(
```

Comment:
can we make this upgradeable?

I'm thinking that it contains storage, and if we need to fix something we can upgrade in place